src/command/repo/get.c

```diff
@@ -46,13 +49,81 @@ storageGetProcess(IoWrite *destination)
 
             if (repoCipherType != cipherTypeNone)
             {
+                /*******************************************************************************************************************
+                Repository encryption design:
+                REPO / (repoCipherPass)
+                     / archive   / (repoCipherPass)
+                     / archive   / stanza / (archiveCipherPass)
+                     / backup    / (repoCipherPass)
+                     / backup    / stanza / (backupCipherPass)
+                     / backup    / stanza / set / (manifestCipherPass)
+                     / backup    / stanza / backup.history / (manifestCipherPass)
+                *******************************************************************************************************************/
+
                 // Check for a passphrase parameter
                 const String *cipherPass = cfgOptionStrNull(cfgOptCipherPass);
 
-                // If not passed as a parameter use the repo passphrase
+                // If not passed as a parameter, find the passphrase
                 if (cipherPass == NULL)
+                {
                     cipherPass = cfgOptionStr(cfgOptRepoCipherPass);
 
+                    const String *relativeFile = strDup(file);
+                    if (strBeginsWith(relativeFile, FSLASH_STR))
+                        relativeFile = strSub(file, strSize(cfgOptionStr(cfgOptRepoPath)) + 1);
```

Comment:
I think we need to check that an absolute path really does being with cfgOptRepoPath. We might get some pretty odd behavior here if it does not.

Reply:
Isn't it the case with https://github.com/pgbackrest/pgbackrest/blob/master/src/storage/storage.c#L703 already ?

Reply:
> Isn't it the case with https://github.com/pgbackrest/pgbackrest/blob/master/src/storage/storage.c#L703 already?

Yes, but that's an assert error which a user should never see. The assert in storagePath() is the last resort if user-provided data is not properly validated.

Reply:
Ok. Let's throw a FormatError also then ?

```
if (!strBeginsWith(relativeFile, cfgOptionStr(cfgOptRepoPath)))
{
    THROW_FMT(
        FormatError, "absolute path '%s' is not in base path '%s'", strPtr(relativeFile),
        strPtr(cfgOptionStr(cfgOptRepoPath)));
}
```

Reply:
Actually, let's use `OptionInvalidValueError` for both these errors. Since this is invalid user input I think that's more appropriate.

Reply:
Seems better indeed. I've updated the 3rd Error regarding stanza not found which is the same type of error.

Reply:
I think what we need is to move this validation code up to line 33 so it happens before storageNewRead(). So, something like (not tested!):
```
    if (strLstSize(cfgCommandParam()) == 1)
    {
        file = strLstGet(cfgCommandParam(), 0);

        if (strBeginsWith(relativeFile, FSLASH_STR))
        {
            if (!strBeginsWith(relativeFile, cfgOptionStr(cfgOptRepoPath)))
            {
                THROW_FMT(
                    OptionInvalidValueError, "absolute path '%s' is not in base path '%s'", strPtr(relativeFile),
                    strPtr(cfgOptionStr(cfgOptRepoPath)));
            }

            file = strSub(file, strSize(cfgOptionStr(cfgOptRepoPath)) + 1);
        }
    }
    else
        THROW(ParamRequiredError, "source file required");
```
Then we don't need the "relativeFile" var anymore. This validation needs to be done regardless of encryption so it makes sense to move it to the beginning.

Reply:
So, yeah, we were getting an assert before, but repo-get has been an internal command used only for testing. If we want it to be public-facing we need to make it cleaner.

Reply:
Indeed, good catch.

I've modified the code accordingly and added extra verifications:
* nothing in repo root directory
* only archive/backup subdirectories allowed

Still the possibility to get unencrypted files and files encrypted with custom `--cipher-pass`.

The tests have been modified accordingly and coverage reports `INFO: tested modules have full coverage`.

Reply:
Hm, github mystery. I pushed a new commit appeared in the fork, not here :(

Reply:
I also didn't get an email notification of your last comment. Maybe wait a bit and see if github catches up?

Reply:
Doesn't matter either way. I can see the commits on your repo so I'll pull those down to do my final review.